### PR TITLE
Improve error message for ImageSpec

### PR DESCRIPTION
**Pull request by pingsutw**
## Tracking issue
NA

## Why are the changes needed?
To improve the error message

### Setup process
```python
from flytekit import task, workflow, ImageSpec

scraper_image_spec3 = ImageSpec(
    base_image="python:3.11-slim-bookworm",
    packages=["pandas", "flytekit"],
    apt_packages=["vim"],
    registry="pingsutw",
)

@task(container_image=scraper_image_spec3)
def t3(a: int) -> int:
    return a + 2


@workflow
def wf(a: int = 3):
    t3(a=a)
```

### Screenshots

Before:
image

After:

Screenshot 2024-06-18 at 2 22 33 PM
Screenshot 2024-06-18 at 2 02 33 PM
Screenshot 2024-06-18 at 2 03 06 PM

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
NA

## Docs link
NA

### CODE REVIEW

File: flytekit/image_spec/image_spec.py [hunk 1]
```diff
@@ -153,9 +158,9 @@ def exist(self) -> bool:
                     f"    pip install --upgrade docker"
                 )
 
-            click.secho(f"Failed to check if the image exists with error : {e}", fg="red")
-            click.secho("Flytekit assumes that the image already exists.", fg="blue")
-            return True
+            click.secho(f"Failed to check if the image exists with error:\n {e}", fg="red")
+            click.secho(f"Flytekit assumes the image {self.image_name()} already exists.", fg="blue")
```
Comment (thomasjpfan) [inline flytekit/image_spec/image_spec.py:162]: I think this print statement belongs in `should_build` instead.

File: flytekit/image_spec/image_spec.py [hunk 2]
```diff
@@ -121,7 +123,10 @@ def exist(self) -> bool:
         except APIError as e:
             if e.response.status_code == 404:
                 return False
-            return True
+
+            click.secho(f"Failed to check if the image exists with error:\n {e}", fg="red")
+            click.secho(f"Flytekit assumes the image {self.image_name()} already exists.", fg="blue")
```
Comment (thomasjpfan) [inline flytekit/image_spec/image_spec.py:128]: Same here

File: flytekit/image_spec/image_spec.py [hunk 3]
```diff
@@ -107,6 +107,8 @@ def is_container(self) -> bool:
     def exist(self) -> bool:
         """
         Check if the image exists in the registry.
+        Return True if the image exists in the registry, False otherwise.
+        Return None if failed to check if the image exists due to the permission issue or other reasons.
```
Comment (thomasjpfan) [inline flytekit/image_spec/image_spec.py:111]: That Python typing for `exist` needs to be updated to include `None`.

File: flytekit/image_spec/image_spec.py [hunk 4]
```diff
@@ -153,9 +158,9 @@ def exist(self) -> bool:
                     f"    pip install --upgrade docker"
                 )
 
-            click.secho(f"Failed to check if the image exists with error : {e}", fg="red")
-            click.secho("Flytekit assumes that the image already exists.", fg="blue")
```
Comment (kumare3) [inline flytekit/image_spec/image_spec.py:157]: Can we ensure we handle runtime error nicely in the clean exception in pyflyte

File: plugins/flytekit-envd/flytekitplugins/envd/image_builder.py [hunk 5]
```diff
@@ -68,7 +80,7 @@ def execute_command(command: str):
 
     if p.returncode != 0:
         _, stderr = p.communicate()
-        raise Exception(f"failed to run command {command} with error {stderr}")
+        raise Exception(f"failed to run command {command} with error:\n {stderr.decode()}")
```
Comment (kumare3) [inline plugins/flytekit-envd/flytekitplugins/envd/image_builder.py:83]: Change this to raise assertionerror
Or runtime or specific error

File: flytekit/image_spec/image_spec.py (hunk 1 above)
Comment (pingsutw) [reply flytekit/image_spec/image_spec.py]: That makes sense, I updated it, thanks!

File: flytekit/image_spec/image_spec.py (hunk 2 above)
Comment (pingsutw) [reply flytekit/image_spec/image_spec.py]: done

File: flytekit/image_spec/image_spec.py (hunk 3 above)
Comment (pingsutw) [reply flytekit/image_spec/image_spec.py:111]: updated

File: flytekit/image_spec/image_spec.py (hunk 4 above)
Comment (pingsutw) [reply flytekit/image_spec/image_spec.py:157]: updated

File: plugins/flytekit-envd/flytekitplugins/envd/image_builder.py (hunk 5 above)
Comment (pingsutw) [reply plugins/flytekit-envd/flytekitplugins/envd/image_builder.py]: updated

**Review by thomasjpfan — APPROVED** [review]

**Pull request merged**